rts/Rendering/Units/UnitDrawer.cpp

```diff
@@ -1333,6 +1329,182 @@ bool CUnitDrawerLegacy::ShowUnitBuildSquare(const BuildInfo& buildInfo, const st
+
+	auto buildCacheItem = std::find_if(buildCache.begin(), buildCache.end(), [hashKey](const BuildCache& bc) {
+		return bc.key == hashKey;
+	});
+
+	if (buildCacheItem == buildCache.end()) {
+		buildCache.emplace_back();
+		buildCacheItem = buildCache.end()-1;
+
+		buildCacheItem->key = hashKey;
+		buildCacheItem->createFrame = gs->frameNum;
+		buildCacheItem->canBuild = CGameHelper::TestUnitBuildSquare(
+			buildInfo,
+			feature,
+			-1,
+			false,
+			&squares[0],
+			&squares[1],
+			&squares[2],
+			&emptyCommands
+		);
+		size_t size = 0;
+		for (int index=0, state=0; state<3; state++) {
+			size += squares[state].size();
+		}
+		buildCacheItem->squares.resize(size);
+		for (int index=0, state=0; state<3; state++) {
+			for (float3 pos: squares[state]) {
+				buildCacheItem->squares[index] = BuildCacheSquare{
+					pos: pos,
+					state: state,
+				};
+				index++;
+			}
+		}
+	}
+
+	bool canBuild = buildCacheItem->canBuild;
+	SColor stateColors[3];
+	if (buildCacheItem->canBuild && !opts.unbuildable) {
+		stateColors[0] =
+			!IsZeroColor(opts.buildableColor) ? opts.buildableColor :
+			!IsZeroColor(opts.defaultColor) ? opts.defaultColor :
+			SColor(cmdColors.buildableSquare);
+	} else {
+		stateColors[0] =
+			!IsZeroColor(opts.unbuildableColor) ? opts.unbuildableColor :
+			!IsZeroColor(opts.defaultColor) ? opts.defaultColor :
+			SColor(cmdColors.unbuildableSquare);
+	}
+	stateColors[1] =
+		!IsZeroColor(opts.featureColor) ? opts.featureColor :
+		!IsZeroColor(opts.defaultColor) ? opts.defaultColor :
+		SColor(cmdColors.featureSquare);
+	stateColors[2] =
+		!IsZeroColor(opts.illegalColor) ? opts.illegalColor :
+		!IsZeroColor(opts.defaultColor) ? opts.defaultColor :
+		SColor(cmdColors.illegalSquare);
+
+	for (auto square: buildCacheItem->squares) {
+		luaBuildSquares.push_back(LuaBuildSquare{
+			pos: square.pos,
+			color: stateColors[square.state],
+		});
+	}
+}
+
+void CUnitDrawerLegacy::ShowLuaBuildSquare() {
+	if (luaBuildSquares.empty() && luaBuildPos.empty()) {
+		return;
+	}
+
+	RECOIL_DETAILED_TRACY_ZONE;
+	glDisable(GL_DEPTH_TEST);
+	glEnable(GL_BLEND);
+	glBlendFunc(GL_SRC_ALPHA, GL_ONE_MINUS_SRC_ALPHA);
+	glDisable(GL_TEXTURE_2D);
+	//glPolygonMode(GL_FRONT_AND_BACK, GL_LINE);
+
+	static auto& rb = RenderBuffer::GetTypedRenderBuffer<VA_TYPE_C>();
+	rb.AssertSubmission();
+
+	auto& sh = rb.GetShader();
+
+	sh.Enable();
+
+	for (const auto &square: luaBuildSquares) {
+		rb.AddQuadLines(
+			{ square.pos                                      , square.color },
+			{ square.pos + float3(SQUARE_SIZE, 0, 0          ), square.color },
+			{ square.pos + float3(SQUARE_SIZE, 0, SQUARE_SIZE), square.color },
+			{ square.pos + float3(0          , 0, SQUARE_SIZE), square.color }
+		);
+	}
+	rb.Submit(GL_LINES);
+
+	SColor startColor(cmdColors.undergroundStart);
+	SColor endColor(cmdColors.undergroundEnd);
+
+	for (int i=0; i<luaBuildPos.size(); i+=2) {
+		float3 pos1 = luaBuildPos[i];
+		float3 pos2 = luaBuildPos[i+1];
```

Comment:
A code comment on why you're consuming 2 at a time and what happens if the number of entries in the table is not a multiple of 2 would be good

Reply:
commented